ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+streaming data.
+
+During research and development, this requirement was translated into a [watch][ClientGoWatch]
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][EventLoopExample].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][EventLoopUsage].
+
+### Upload Mechanism
+
+This EP describes the direct data streaming to the Kubernetes cluster, directly to the POD running
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+The proposed API changes are the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+```
+
+The Build Controller will inspect for a type `LocalCopy` on the `"sources"` slice, and therefore
+instead of adding the git clone step, it will instead wait for the user to upload.
+
+The `sources` slice is part of [Remote Artifacts (EP)][RemoteArtifactsEP] and has been worked on, so
+more effort may be needed to create a good API fit.
+
+In order to let each EP implementation stand-alone, an alternative mechanism should take place, by
+creating a simple annotation in the `BuildRun` object the Build Controller can be notified about the
+type of build expected. As per:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  annotations:
+    build.shipwright.io/upload.wait: true
+    build.shipwright.io/upload.wait.timeout: 120
```

Comment:
I do not see a reason for the first one because the type of the source defines already this behavior.

The second setting has value. I suggest to make it strongly-typed in the `sources`  section. The value would be that we can use the duration type (like we do for the Build(Run) timeout property). We can still make it optional with a reasonable and documented default.

Reply:
The reasoning is this [previous comment](https://github.com/shipwright-io/community/pull/11#discussion_r684971517).